### PR TITLE
cp.8 example changed

### PR DESCRIPTION
**Pull request by thschrapel**
A local copy of a global was used to prevent a data race, but it was uses without validation. The beautiful syntax of the sample distracts from the core problem to solve.

### CODE REVIEW

**Comment by cubbimew** [comment]
This appears to be unnecessary, especially in the context of the rule in question.

**Comment by thschrapel** [comment]
> This appears to be unnecessary, especially in the context of the rule in question.

I remember, some  time ago I read a sentence that reinterpret_cast<> and static_cast<> were designed to look ugly compared to the c-style cast to catch the programmers attention. Furthermore there is a lot of guidance to use explicit conversions like string::c_str(). So if I use an atomic operation, I normally have to reason about the code some time. And then:  why not making that explicit in code instead of writing a comment?

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -14316,14 +14316,30 @@ This has no effect on synchronization: The data race is still there!
...
+    {
+        // will throw on error 
+        // ...
+    }
     Pool* use()
     {
-        if (int n = free_slots--) return &pool[n];
+        // Apply the Copy-Validate-Use strategy:
+        // 1. Atomic substraction + local copy:
+        int n = free_slots.fetch_sub(1); // Use atomic operations explicitly
```
Comment (bgloyer) [inline CppCoreGuidelines.md:14328]: The rule is for ```volatile``` so maybe the details of atomic is out of scope?

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -14316,14 +14316,30 @@ This has no effect on synchronization: The data race is still there!
...
+        // 2. Validate n: 
+        check(n);
```
Comment (bgloyer) [inline CppCoreGuidelines.md:14330]: The ```check()``` is a huge improvement.  The current example doesn't return anything for the ```else``` case and the caller has no way to know.

Maybe more contex could be added to the example too?  Is ```use()``` in a class?  It seems odd to have a function named 'use' that takes no parameters and returns a pointer.

**Comment by bgloyer** [comment]
this is cp.8

File: CppCoreGuidelines.md (hunk 2 above)
Comment (thschrapel) [reply CppCoreGuidelines.md:14330]: Thank you for the positive feedback. 

> The `check()` is a huge improvement. The current example doesn't return anything for the `else` case and the caller has no way to know.
> 
> Maybe more contex could be added to the example too? Is `use()` in a class? It seems odd to have a function named 'use' that takes no parameters and returns a pointer.

Thank you for the positive feedback. I just wanted to add some community feedback so the editors (or other cpp-masters) can review  the 'door-opener' examples of the CP-section. A returning pointer really distracts, maybe just a 'simple' cout << would do , so one does not have to reason about 'dangling' pointers. The 'Every-Stack-Has-Its-Own-Stack' feature is much more worthy to think about especially in terms of thread-safety of (recursive) functions and the implications on resource consumption of threads.

**Comment by cubbimew** [comment]
> why not making that explicit in code instead of writing a comment?

It seems you're proposing a rule then, something like "For atomic types, prefer functions to overloaded operators". I'd probably vote for such rule especially for inc/dec operators vs fetch_add/sub. But a fragment in an  example on the rule about misuse of volatile wouldn't do it justice.

**Comment by BenjamenMeyer** [comment]
> why not making that explicit in code instead of writing a comment?

because comments carry intent; where code only shows what function.
You need both - code and comment - for long term maintenance.

**Comment by thschrapel** [comment]
> > why not making that explicit in code instead of writing a comment?
> 
> It seems you're proposing a rule then, something like "For atomic types, prefer functions to overloaded operators". I'd probably vote for such rule especially for inc/dec operators vs fetch_add/sub. But a fragment in an example on the rule about misuse of volatile wouldn't do it justice.

I like the rule you made from the comment. Here is a good reason for it: Matthew Kulukundis on Cppcon 2021 demonstrated the motivation and the pitfalls behind lock-free programming. I think it's a detailed example for being explicit using 'atomic'/lock-free functionality.

**Comment by hsutter** [comment]
Editors call: Perhaps open an issue regarding whether to consider guidance about using atomic operators vs. named function calls, but this example doesn't seem to benefit from having that discussed added here.

**Pull request closed without merging**